src/sync/chase_lev.rs

```diff
@@ -158,68 +159,82 @@ impl<T> Deque<T> {
+        self.bottom.store(b + 1, Relaxed);
     }
 
     unsafe fn try_pop(&self) -> Option<T> {
         let guard = epoch::pin();
 
-        let b = self.bottom.load(SeqCst);
-        let a = self.array.load(SeqCst, &guard).unwrap();
-        let b = b - 1;
-        self.bottom.store(b, SeqCst);
-        let t = self.top.load(SeqCst);
+        let b = self.bottom.load(Relaxed) - 1;
+        let a = self.array.load(Relaxed, &guard).unwrap();
+        self.bottom.store(b, Relaxed);
+        fence(SeqCst); // the store to bottom must occur before loading top.
+        let t = self.top.load(Relaxed);
+        
         let size = b - t;
-        if size < 0 {
-            self.bottom.store(t, SeqCst);
-            return None
-        }
-        let data = a.get(b);
-        if size > 0 {
-            self.maybe_shrink(b, t, &guard);
-            return Some(data)
-        }
-        if self.top.compare_and_swap(t, t + 1, SeqCst) == t {
-            self.bottom.store(t + 1, SeqCst);
-            return Some(data)
+        if size >= 0 {
+            // non-empty case
+            let mut data = Some(a.get(b));
+            if size == 0 {
+                // last element in queue, check for races.
+                if self.top.compare_and_swap(t, t + 1, SeqCst) != t {
+                    // lost the race.
+                    mem::forget(data.take());
+                }
+                
+                // set the queue to a canonically empty state.
+                self.bottom.store(b + 1, Relaxed);
+            } else {
+                self.maybe_shrink(b, t, &guard);
+            }
+            data
         } else {
-            self.bottom.store(t + 1, SeqCst);
-            mem::forget(data); // someone else stole this value
-            return None
+            // empty queue. revert the decrement of "b" and try to shrink.
+            self.bottom.store(b + 1, Relaxed);
+            None
         }
     }
 
     fn steal(&self) -> Steal<T> {
         let guard = epoch::pin();
 
-        let t = self.top.load(SeqCst);
-        let old = self.array.load(SeqCst, &guard).unwrap();
-        let b = self.bottom.load(SeqCst);
-        let a = self.array.load(SeqCst, &guard).unwrap();
+        let t = self.top.load(Acquire);
+        let old = self.array.load(Acquire, &guard).unwrap();
+        fence(SeqCst); // top must be loaded before bottom.
+        let b = self.bottom.load(Acquire);
+        let a = self.array.load(Acquire, &guard).unwrap();
         let size = b - t;
         if size <= 0 {
             return Steal::Empty
         }
+        
+        // check for a specific edge case where the buffer has been shrunk while
+        // this steal operation was in progress.
         if (size as usize) % a.size() == 0 {
-            if *a as *const _ == *old as *const _ && t == self.top.load(SeqCst) {
-                return Steal::Empty
+            if *a as *const _ == *old as *const _  && t == self.top.load(Acquire) {
```

Comment:
Ah right! One of these days I may understand this memory model... In that case this is probably fine